# 1. Customer Segmentation Data Cleaning and Preparation

## Introduction

This project focuses on cleaning and preparing a dataset for customer segmentation analysis. The dataset includes demographic, behavioral, and preference information for over 53,000 customers. Properly cleaning the data is essential for accurate segmentation and actionable insights. The cleaned dataset will be used for advanced customer analysis in future projects.

### Objectives

1. **Data Cleaning**:
   - Remove duplicate entries to ensure data integrity.
   - Convert relevant columns (e.g., `Purchase History`) to appropriate formats for analysis.
   - Handle missing values effectively.

2. **Data Inspection**:
   - Verify column types and data completeness.
   - Summarize key numerical variables to understand their distributions.

3. **Dataset Preparation**:
   - Save the cleaned dataset for segmentation modeling and exploratory analysis.

---

In [2]:
import pandas as pd
import numpy as np
df = pd.read_csv(r"C:\Users\Zana\Desktop\portfolio_projects\project_2\customer_segmentation_data.csv")
df.head()

Customer ID  Age  Gender Marital Status    Education Level  \
0        84966   23  Female        Married   Associate Degree   
1        95568   26    Male        Widowed          Doctorate   
2        10544   29  Female         Single   Associate Degree   
3        77033   20    Male       Divorced  Bachelor's Degree   
4        88160   25  Female      Separated  Bachelor's Degree   

  Geographic Information    Occupation  Income Level Behavioral Data  \
0                Mizoram  Entrepreneur         70541         policy5   
1                    Goa       Manager         54168         policy5   
2              Rajasthan  Entrepreneur         73899         policy5   
3                 Sikkim  Entrepreneur         63381         policy5   
4            West Bengal       Manager         38794         policy1   

  Purchase History Interactions with Customer Service  \
0       04-10-2018                              Phone   
1       11-06-2018                               Chat   
2       06-05-2021                              Email   
3       09-02-2018                               Chat   
4       09-10-2018                               Chat   

  Insurance Products Owned  Coverage Amount  Premium Amount Policy Type  \
0                  policy2           366603            2749       Group   
1                  policy1           780236            1966       Group   
2                  policy3           773926            4413       Group   
3                  policy2           787815            4342      Family   
4                  policy4           366506            1276      Family   

  Customer Preferences Preferred Communication Channel Preferred Contact Time  \
0                Email               In-Person Meeting              Afternoon   
1                 Mail               In-Person Meeting                Morning   
2                Email                            Mail                Evening   
3                 Text               In-Person Meeting                Anytime   
4                Email                            Text               Weekends   

  Preferred Language Segmentation Group  
0            English           Segment5  
1             French           Segment5  
2             German           Segment3  
3             French           Segment3  
4            English           Segment2

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53503 entries, 0 to 53502
Data columns (total 20 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Customer ID                         53503 non-null  int64 
 1   Age                                 53503 non-null  int64 
 2   Gender                              53503 non-null  object
 3   Marital Status                      53503 non-null  object
 4   Education Level                     53503 non-null  object
 5   Geographic Information              53503 non-null  object
 6   Occupation                          53503 non-null  object
 7   Income Level                        53503 non-null  int64 
 8   Behavioral Data                     53503 non-null  object
 9   Purchase History                    53503 non-null  object
 10  Interactions with Customer Service  53503 non-null  object
 11  Insurance Products Owned            53503 non-null  ob

In [4]:
df.describe()

Customer ID           Age   Income Level  Coverage Amount  \
count   53503.000000  53503.000000   53503.000000     53503.000000   
mean    52265.204998     44.140945   82768.324318    492580.789638   
std     28165.000067     15.079486   36651.075670    268405.505571   
min         1.000000     18.000000   20001.000000     50001.000000   
25%     28950.500000     32.000000   51568.500000    249613.500000   
50%     55858.000000     43.000000   80719.000000    477261.000000   
75%     76096.000000     57.000000  115973.500000    739124.000000   
max    100000.000000     70.000000  149999.000000   1000000.000000   

       Premium Amount  
count    53503.000000  
mean      3023.702447  
std       1285.834295  
min        500.000000  
25%       1817.000000  
50%       3194.000000  
75%       4311.500000  
max       5000.000000

In [6]:
# Deleted Duplicates
df_cleaned = df.drop_duplicates()

In [7]:
# Convert Purchase History to datetime format
df_cleaned['Purchase History'] = pd.to_datetime(df_cleaned['Purchase History'], errors='coerce')

In [8]:
# Check for missing values
print(df_cleaned.isnull().sum())

Customer ID                               0
Age                                       0
Gender                                    0
Marital Status                            0
Education Level                           0
Geographic Information                    0
Occupation                                0
Income Level                              0
Behavioral Data                           0
Purchase History                      30808
Interactions with Customer Service        0
Insurance Products Owned                  0
Coverage Amount                           0
Premium Amount                            0
Policy Type                               0
Customer Preferences                      0
Preferred Communication Channel           0
Preferred Contact Time                    0
Preferred Language                        0
Segmentation Group                        0
dtype: int64


In [10]:
# Replaced missing values in "Purchase History" with NaN
df_cleaned['Purchase History'] = df_cleaned['Purchase History'].fillna(np.nan)

# Checked missing value
print(df_cleaned.isnull().sum())

# Display a sample of the cleaned data to see changes
print(df_cleaned.head())

Customer ID                               0
Age                                       0
Gender                                    0
Marital Status                            0
Education Level                           0
Geographic Information                    0
Occupation                                0
Income Level                              0
Behavioral Data                           0
Purchase History                      30808
Interactions with Customer Service        0
Insurance Products Owned                  0
Coverage Amount                           0
Premium Amount                            0
Policy Type                               0
Customer Preferences                      0
Preferred Communication Channel           0
Preferred Contact Time                    0
Preferred Language                        0
Segmentation Group                        0
dtype: int64
   Customer ID  Age  Gender Marital Status    Education Level  \
0        84966   23  Female        Married

In [11]:
# Save the cleaned DataFrame to a CSV file
df_cleaned.to_csv('cleaned_customer_segmentation_data.csv', index=False)

## Key Steps and Results

1. **Data Cleaning**:
   - Removed duplicate rows from the dataset, resulting in a unique and reliable dataset.
   - Converted the `Purchase History` column to datetime format for temporal analysis.
   - Replaced missing values in the `Purchase History` column with `NaN` for future handling.

2. **Inspection of Missing Values**:
   - Identified missing values in the `Purchase History` column, with **30,808 missing entries**.
   - Other columns were complete, ensuring a robust dataset for segmentation.

3. **Descriptive Statistics**:
   - Key numerical columns (`Age`, `Income Level`, `Coverage Amount`, and `Premium Amount`) were summarized:
     - **Age**: Customers range from 18 to 70 years, with an average age of **44**.
     - **Income Level**: Averages around **82,768**, with values spanning from **20,001** to **149,999**.
     - **Coverage Amount**: Customers have varying coverage amounts, averaging **492,580**.
     - **Premium Amount**: Premiums average **3,023**, indicating diverse policy pricing.

4. **Final Dataset**:
   - Cleaned dataset includes **53,503 unique customers** with 20 columns.
   - The cleaned dataset was saved as **`cleaned_customer_segmentation_data.csv`** for further analysis.

---